File: qbopt/frontend/raising_numeric_policy.py
"""Separate BASIC's explicit overflow observation from native integer arithmetic."""

from dataclasses import replace

from iced_x86 import Code

from qbopt.model import ir
from qbopt.model import mir
from qbopt.model.floating import Exceptions
from qbopt.frontend.raising_calls import _discarded


def checkpoints(body: mir.MirBody) -> mir.MirBody:
    return replace(
        body,
        blocks=tuple(
            replace(
                block,
                ops=tuple(
                    _discarded(op)
                    if op.kind is mir.Kind.FCHECK
                    else replace(op, floating=replace(op.floating, exceptions=Exceptions.DEFERRED))
                    if op.floating is not None
                    else op
                    for op in block.ops
                ),
            )
            for block in body.blocks
        ),
    )


def native(body):
    return replace(
        body,
        blocks=tuple(
            replace(
                block,
                ops=tuple(
                    _discarded(op) if isinstance((node := getattr(op, "node", None)), ir.Opaque)
                    and node.insn.insn.code == Code.INTO else op
                    for op in block.ops
                ),
            )
            for block in body.blocks
        ),
    )

File: src/frontend/raising_numeric_policy.rs
//! Port of `qbopt/frontend/raising_numeric_policy.py`: separate BASIC's
//! explicit overflow observation from native integer arithmetic.

use iced_x86::Code;

use crate::frontend::raising_calls::_discarded;
use crate::model::floating::Exceptions;
use crate::model::ir::nodes::Node;
use crate::model::mir::{Kind, RaisedBody};

pub fn checkpoints(mut body: RaisedBody) -> RaisedBody {
    for block in &mut body.body_mut().blocks {
        for op in &mut block.ops {
            if op.kind == Kind::Fcheck {
                *op = _discarded(op);
            } else if let Some(floating) = &mut op.floating {
                floating.exceptions = Exceptions::Deferred;
            }
        }
    }
    body
}

pub fn native(mut body: RaisedBody) -> RaisedBody {
    for block in &mut body.body_mut().blocks {
        for op in &mut block.ops {
            if matches!(op.node().map(|node| &**node), Some(Node::Opaque(node)) if node.insn.insn.code() == Code::Into) {
                *op = _discarded(op);
            }
        }
    }
    body
}

#[cfg(test)]
#[path = "raising_numeric_policy_tests.rs"]
mod tests;
